src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +147,109 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
+
+            working_dir = step_exec.get("WorkingDirectory")
+            run_script = validate_run_step(parsed_obj)
+            if run_script is None:
+                raise CallGraphError(f"Invalid run step at {source_path}.")
+            parsed_content = parse(run_script, macaron_path=macaron_path)
+        case BashScriptType.FILE:
+            parsed_content = parse_file(source_path, macaron_path=macaron_path)
+    bash_node = BashNode(
+        name,
+        node_type,
+        source_path,
+        parsed_step_obj=parsed_obj,
+        parsed_bash_obj=parsed_content,
+        node_id=node_id,
+        caller=caller,
+    )
+    caller_commands = parsed_content.get("commands", [])
 
     # Parse the bash script files called from the current script.
-    if recursive and repo_path:
+    if caller_commands and repo_path:
         for cmd in caller_commands:
             # Parse the scripts that end with `.sh`.
-            # We only parse recursively at depth 1, so don't set the recursive argument in parse_file().
             # TODO: parse Makefiles for bash commands.
-            if cmd[0] and cmd[0].endswith(".sh") and os.path.exists(os.path.join(repo_path, cmd[0])):
-                callee_commands = parse_file(os.path.join(repo_path, cmd[0])).get("commands", [])
-                if not callee_commands:
-                    continue
-
-                yield BashCommands(
-                    caller_path=os.path.join(working_dir, cmd[0]),
-                    CI_path=ci_file,
-                    CI_type=ci_type,
-                    commands=callee_commands,
-                    job_name=job_name,
-                    step_name=step_name,
-                )
+            if not cmd[0] or not cmd[0].endswith(".sh"):
+                continue
+
+            # Check for path traversal patterns before analyzing a bash file.
+            bash_file_path = os.path.realpath(os.path.join(repo_path, working_dir or "", cmd[0]))
+            if os.path.exists(bash_file_path) and bash_file_path.startswith(repo_path):
```

Comment:
Hmm is `repo_path` in its canonical form at this point of code. I think it would be safer if we also convert `repo_path` to its canonical form using `os.path.realpath` before performing the comparison `bash_file_path.startswith(repo_path)`.

Reply:
Why shouldn't it be canonical? `repo_path` is what Macaron itself creates, right? How can it be non-canonical?

Reply:
You are correct. This `repo_path` is obtained from `Repository.fs_path` (https://github.com/oracle/macaron/blob/db3231ffe0334c49f78582f6945ac2996087bc99/src/macaron/database/table_definitions.py#L296) which has been resolved to canonical form by:
- gitpython (when we are analyzing a remote path repository) - https://github.com/oracle/macaron/blob/10658c5d2f31d0e27fce4d39f800d21e39c24ceb/src/macaron/slsa_analyzer/analyzer.py#L447
- this method when we are analyzing a local repository - https://github.com/oracle/macaron/blob/10658c5d2f31d0e27fce4d39f800d21e39c24ceb/src/macaron/slsa_analyzer/analyzer.py#L863

Reply:
I don't think we need to do anything else regarding this comment.